### DISCUSSION TITLE
Component 'entity_linker' could not be run. Did you forget to call `initialize()`

### DISCUSSION THREAD

**Discussion by amirj** [explosion/spaCy#8080] [Help: Coding & Implementations] [answered]
I can train an entity linker and perfectly use it! :-) 
Now, I'm going to overwrite the default "kb.get_alias_candidates" to return a custom list of candidates:

```python
@registry.misc("custom_candidate_generator")
def create_candidates():
    return get_custom_candidates


def get_custom_candidates(kb, span):
    candidates = []
    for entity_id, score in candidate_generator_func(span):
        candidate = Candidate(
            kb=kb,
            entity_hash=kb.vocab.strings[entity_id],
            entity_freq=111,
            entity_vector=kb.get_vector(entity_id),
            alias_hash=kb.vocab.strings[span.text],
            prior_prob=score,
        )
        candidates.append(candidate)
    return candidates
```

`candidate_generator_func` is a function that takes as input an alias/span and returns a list of entity_ids/scores. `entity_id` is a string similar to what I used to [add_entity to the knowledgebase](https://spacy.io/api/kb#add_entity). After adding the above custom candidate generator, the entity linker model is trained as before but when calling `nlp(text)`, I'm getting the following error:

```
/home/user/.conda/envs/sample-entity-linker/lib/python3.8/site-packages/spacy/pipeline/entity_linker.py:362: RuntimeWarning: invalid value encountered in true_divide
  sims = xp.dot(
Traceback (most recent call last):
  File "/home/user/.conda/envs/sample-entity-linker/lib/python3.8/site-packages/spacy/language.py", line 995, in __call__
    doc = proc(doc, **component_cfg.get(name, {}))
  File "spacy/pipeline/trainable_pipe.pyx", line 56, in spacy.pipeline.trainable_pipe.TrainablePipe.__call__
  File "/home/user/.conda/envs/sample-entity-linker/lib/python3.8/site-packages/spacy/util.py", line 1503, in raise_error
    raise e
  File "spacy/pipeline/trainable_pipe.pyx", line 52, in spacy.pipeline.trainable_pipe.TrainablePipe.__call__
  File "/home/user/.conda/envs/sample-entity-linker/lib/python3.8/site-packages/spacy/pipeline/entity_linker.py", line 373, in predict
    final_kb_ids.append(best_candidate.entity_)
  File "spacy/kb.pyx", line 45, in spacy.kb.Candidate.entity_.__get__
  File "spacy/strings.pyx", line 132, in spacy.strings.StringStore.__getitem__
KeyError: "[E018] Can't retrieve string for hash '16817217201437651034'. This usually refers to an issue with the `Vocab` or `StringStore`."

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "sample_entity_linker/models/spacy_entity_linker.py", line 296, in <module>
    for ent in spacy_entity_linker.map_to_entities(text):
  File "sample_entity_linker/models/spacy_entity_linker.py", line 228, in map_to_entities
    doc = self.nlp(text)
  File "/home/user/.conda/envs/sample-entity-linker/lib/python3.8/site-packages/spacy/language.py", line 998, in __call__
    raise ValueError(Errors.E109.format(name=name)) from e
ValueError: [E109] Component 'entity_linker' could not be run. Did you forget to call `initialize()`?
```

What I missed in my custom candidate generator function?!

**Comment by amirj** [comment, marked as answer]
Describing the problem helped me to find the bug! --some entity-ids was not available in KB for some reasons.

Now, my question is that why I didn't get this error during the training?! How can I make sure that entity linker use the custom candidate generation during the training phase, especially the prior probabilities (I set "incl_prior" to True when creating entity_linker pipe)?

**Reply by svlandeg** [reply]
Hi @amirj,

> some entity-ids was not available in KB for some reasons

I'm not sure why this happens. If you're using the built-in `KnowledgeBase`, the method `add_entity` makes sure to add the entity ID to the `vocab`:
```
entity_hash = self.vocab.strings.add(entity)
```
In your `get_custom_candidates`, you probably want to double check that the entity you're returning, is actually in the `KnowledgeBase`:
```
present = kb.contains_entity(entity_id)
```

> Now, my question is that why I didn't get this error during the training?! 

I can't really address this question without having the full code. Why weren't the entities in the KB in the first place? If there were only a few cases, perhaps you just didn't run into those during training?

> How can I make sure that entity linker use the custom candidate generation during the training phase

Every `entity_linker` component is always constructed with a  `get_candidates` function. If you didn't define it specifically, the default one will be used (which is not what you want). You should be able to do something like this:
```
config = {
   "get_candidates": {'@misc': 'custom_candidate_generator'},
}
> nlp.add_pipe("entity_linker", config=config)
```
if you're writing the training loop in code, or equivalently, if you're using the training config:
```
[components.entity_linker]
factory = "entity_linker"
get_candidates = {"@misc":"custom_candidate_generator"}
```
